### PR TITLE
Implement some more extension functions

### PR DESCRIPTION
**Pull request by sbeca**
This PR started from a want to add `Process.WaitForExitAsync`.

The .NET library source for `Process.WaitForExitAsync` uses `CancellationToken.UnsafeRegister` so I also added those into this PR, as well as a missing `CancellationToken.Register` overload that I came across while copying tests over for `CancellationToken.UnsafeRegister`.

Also, while working on the tests for `CancellationToken.UnsafeRegister`, I noticed that `CancellationTokenSource.CancelAsync` had been added in a previous commit, but no tests were added for it at the same time. So I figured I should copy some tests for `CancellationTokenSource.CancelAsync` from the .NET repo while I was working in that area, which led me to hit a deadlock with the current implementation of `CancellationTokenSource.CancelAsync`. So I also added a workaround in that function while I was there.

I didn't expect to go down such a rabbit hole when all I wanted was `Process.WaitForExitAsync`, but I think everything I've added in this PR is worthwhile adding to the library.

### CODE REVIEW

File: src/Tests/BuildApiTest.cs [hunk 1]
```diff
@@ -17,8 +18,8 @@ class BuildApiTest
     public void Run()
     {
         var solutionDirectory = SolutionDirectoryFinder.Find();
-        var path = Path.Combine(solutionDirectory, @"Consume\bin\Debug\netstandard2.0\Consume.dll");
-        var md = Path.Combine(solutionDirectory, @"..\api_list.include.md");
+        var path = Path.Combine(solutionDirectory, "Consume", "bin", "Debug", "netstandard2.0", "Consume.dll");
+        var md = Path.Combine(solutionDirectory, "..", "api_list.include.md");
```
Comment (sbeca) [inline src/Tests/BuildApiTest.cs:22]: This was changed so that I could run it on macOS

**Comment by sbeca** [comment]
I'll fix the build errors now

**Comment by sbeca** [comment]
Build/test errors have been fixed

**Comment by SimonCropp** [comment]
@sbeca  very nice work. will merge and deploy this now. should be on nuget in ~20min

**Pull request merged**